### PR TITLE
Add Skeleton and Skinning components to Radium

### PR DESCRIPTION
**Pull request by nmellado**
WarningL this PR is still work in progress, and should be used to discuss and design the final solution.

* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Be aware that the PR request cannot be accepted if it doesn't pass the Continuous Integration tests.

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Bring functionality from plugin to Radium.

* **What is the current behavior?** (You can also link to an open issue here)
Components and Systems that are necessary to use Skeletons are not part of the Radium library, but rather defined in a plugin https://gitlab.com/Storm-IRIT/radium-official-plugins/skeletonbasedanimation

* **What is the new behavior (if this is a feature change)?**
Add required Components and Systems to Radium.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/Engine/Component/SkeletonComponent.cpp  2
- src/Engine/Component/SkinningComponent.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5394305&bid=17683086)

File: src/Engine/Component/SkinningComponent.cpp [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+        if ( mesh->hasAttrib( attrUV ) )
+        {
+            auto handle = mesh->getAttribHandle<Ra::Core::Vector3>( attrUV );
+            m_baseUV    = mesh->getAttrib( handle ).data();
+        }
+
+        m_weightTechnique.reset( new Ra::Engine::RenderTechnique );
+        auto builder = Ra::Engine::EngineRenderTechniques::getDefaultTechnique( "BlinnPhong" );
+        builder.second( *m_weightTechnique.get(), true );
+        auto matT = std::static_pointer_cast<Ra::Engine::BlinnPhongMaterial>(
```
Comment (nmellado) [inline src/Engine/Component/SkinningComponent.cpp:158]: Codacy Issue found: [Variable 'matT' is assigned a value that is never used.](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5394305)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/Engine/Component/SkinningComponent.cpp  1
- src/Engine/Component/SkeletonComponent.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5394305&bid=17683086)

File: src/Engine/Component/SkinningComponent.cpp (hunk 1 above)
Comment (nmellado) [inline src/Engine/Component/SkinningComponent.cpp:158]: Codacy Issue found: [Variable 'matT' is assigned a value that is never used.](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5394305)

**Comment by nmellado** [comment]
I have moved most of the files to this PR.
What is missing is this file: https://gitlab.com/Storm-IRIT/radium-official-plugins/skeletonbasedanimation/-/blob/master/src/Drawing/SkeletonBoneDrawable.hpp

The class `SkeletonBoneDrawable` does not fit the standard Radium data-structures: it offers methods to generate and manipulate the bones RO, but it has no base class in Radium. 
@dlyr / @MathiasPaulin , what do you think about it ? Do you have any suggestion on how we could handle this ?

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/Engine/Component/SkinningComponent.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5394305&bid=17683086)

**Comment by MathiasPaulin** [comment]
> I have moved most of the files to this PR.
> What is missing is this file: https://gitlab.com/Storm-IRIT/radium-official-plugins/skeletonbasedanimation/-/blob/master/src/Drawing/SkeletonBoneDrawable.hpp
> 
> The class `SkeletonBoneDrawable` does not fit the standard Radium data-structures: it offers methods to generate and manipulate the bones RO, but it has no base class in Radium.
> @dlyr / @MathiasPaulin , what do you think about it ? Do you have any suggestion on how we could handle this ?

I've no idea of what we can do. Where is this classe used ? What it provides ? Is there some documentation around there for this class ? I'll give a look to try to understand what is the meaning of this.

**Comment by MathiasPaulin** [comment]
I gave a look on the whole PR. It will take some time to make it approved ;) . Nevertheless, for the `SkeletonBoneDrawable`, I look for its usage. It's no more than a specific `Ra::Engine::RenderObject` but this class is marked `final` so no inheritance is allowed. Association (what is done by `SkeletonBoneDrawable` is the best we can do. But, the way it is managed in this PR is far from optimal and still relies on old fashioned way of managing things on Radium. When the PR will be ready, I'll try to refactor its whole code to have better Radium compliance and performances. For now, you just have to get the service working, I'll take the following ...

**Comment by nmellado** [comment]
Ok will do it. Thanks

**Comment by MathiasPaulin** [comment]
Is this PR a subset of #551 ? The same files are changed (with 4 more in #551) and the same goal is claimed ? Which one to review/merge ?

**Comment by hoshiryu** [comment]
This PR was introduced to talk about how to nicely internalize skeleton based animation components into Radium Engine.
#551 is one proposal inspired from PRs #539 and #541

**Pull request closed without merging**